rasterio/windows.py

```diff
@@ -207,13 +216,23 @@ def intersection(*windows):
     -------
     Window
     """
-    if not intersect(windows):
-        raise WindowError("windows do not intersect")
-
-    stacked = np.dstack([toranges(w) for w in windows])
-    return Window.from_slices(
-        (stacked[0, 0].max(), stacked[0, 1].min()),
-        (stacked[1, 0].max(), stacked[1, 1]. min()))
+    return functools.reduce(_intersection, windows)
+
+def _compute_intersection(w1, w2):
+    """ Compute intersection of window 1 and window 2"""
+    col_off = max(w1.col_off, w2.col_off)
+    row_off = max(w1.row_off, w2.row_off)
+    width = min(w1.col_off+w1.width, w2.col_off+w2.width) - col_off
+    height = min(w1.row_off+w1.height, w2.row_off+w2.height) - row_off
+    return col_off, row_off, width, height
+
+def _intersection(w1, w2):
+    """ Compute intersection of window 1 and window 2"""
+    coeffs = _compute_intersection(w1, w2)
+    if coeffs[2] > 0 and coeffs[3] > 0:
+        return Window(*coeffs)
+    else:
+        raise WindowError(f"Intersection is empty {w1} {w2}")
```

Comment:
@groutr combine these as the first is only called from the second?

Reply:
I think you're referring to why I split the functions up the way I did.

The reasoning is as follows:
`_compute_intersection`: Compute the intersection directly. A low-level function that is useful in the case where it is not desired to create a Window object. This is useful in the case where the calling method is creating its own Window object or manipulating the values. Reviewing the function, it should probably be modified to also work on the tuple form of a Window.
`_intersection`: Compute an intersection and check if the intersection is empty. It is a higher-level function that is primarily designed to receive as input Windows and return a Window. This is what makes the `reduce` call possible.

I can combine them if you feel the current split is unnecessary.